fix silent correctness bug with channels_last usage of upsample cuda kernels

bdhirsh opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/54590

After the porting the upsample operators to be structured, they now forward memory_format information to the output. This is a problem for the cuda kernels, which are not implemented to deal with `torch.channels_last` memory format. After the structured port, the kernels now expect the output to already be made contiguous if it was not explicitly passed in. The cuda kernel is written to only handle "normal" contiguous tensors, and not channels_last contiguous tensors. The updated operators are:
* upsample_nearest2d
* upsample_nearest3d

It looks like upsample_bilinear2d and upsample_trilinear3d don't have this issue, so I didn't change those kernels.

This fix just allocates a temporary, contiguous output tensor when that happens, writes the results to the temporary and copies the results back to the output tensor.


Also added tests for upsample_nearest2d/3d to explicitly check that cuda and cpu results are the same. Those tests need to have input tensors with a mix of values and a large enough dimension size to distinguish between contiguous and channels_last behavior.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54744 fix silent correctness bug with channels_last usage of upsample cuda kernels**

Differential Revision: [D27351393](https://our.internmc.facebook.com/intern/diff/D27351393)

bdhirsh committed:
fix silent correctness bug with channels_last usage of upsample cuda kernels

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/cuda/UpSampleBilinear2d.cu:
```diff
@@ -179,7 +184,7 @@ static void upsample_bilinear2d_out_cuda_template(
 
         auto idata = input.packed_accessor64<scalar_t, 4>();
-        auto odata = output.packed_accessor64<scalar_t, 4>();
+        auto odata = output_c.packed_accessor64<scalar_t, 4>();
```
It looks like `packed_accessor` correctly handles the strides of the tensor, and bilinear/trilinear interpolation produce correct results even without this fix, so maybe undo these changes?
Difference of 1e-4 looks kinda large, but maybe that's and in any case it's not changed by this

ngimel reviewed (commented):
Cool, please add a test!

bdhirsh committed:
Update on "fix silent correctness bug with channels_last usage of upsample cuda kernels"

Fixes https://github.com/pytorch/pytorch/issues/54590

After the porting the upsample operators to be structured, they now forward memory_format information to the output. This is a problem for the cuda kernels, which are not implemented to deal with `torch.channels_last` memory format. The operators are:
* upsample_nearest2d
* upsample_bilinear2d
* upsample_nearest3d
* upsample_trilinear3d

This fix just allocates a temporary, contiguous output tensor when that happens, writes the results to the temporary and copies the results back to the output tensor.

I held off on adding tests to get the fix out quickly, but I wrote a script and ran some manual tests, that basically just asserts that the outputs are the same for cpu and cuda, for some threshold. I ran it for all 4 operators:
```
import torch

def basically_equal(t1, t2):
    epsilon = 1e-4
    diffs = torch.abs(t1 - t2)
    print(torch.all(diffs < 1e-4))

# upsample 2d
a = torch.arange(48).reshape(2, 2, 3, 4).contiguous(memory_format=torch.channels_last).float()

out_cpu = torch.nn.functional.interpolate(a, scale_factor=2, mode='nearest')
out_cuda = torch.nn.functional.interpolate(a.to('cuda'), scale_factor=2, mode='nearest')

basically_equal(out_cpu, out_cuda.to("cpu"))

out_cpu = torch.nn.functional.interpolate(a, scale_factor=2, mode='bilinear', align_corners=True)
out_cuda = torch.nn.functional.interpolate(a.to('cuda'), scale_factor=2, mode='bilinear', align_corners=True)

basically_equal(out_cpu, out_cuda.to("cpu"))

# upsample 3d
a = torch.arange(96).reshape(2, 2, 2, 3, 4).contiguous(memory_format=torch.channels_last_3d).float()

out_cpu = torch.nn.functional.interpolate(a, scale_factor=3, mode='nearest')
out_cuda = torch.nn.functional.interpolate(a.to('cuda'), scale_factor=3, mode='nearest')

basically_equal(out_cpu, out_cuda.to("cpu"))

out_cpu = torch.nn.functional.interpolate(a, scale_factor=3, mode='trilinear', align_corners=True)
out_cuda = torch.nn.functional.interpolate(a.to('cuda'), scale_factor=3, mode='trilinear', align_corners=True)

basically_equal(out_cpu, out_cuda.to("cpu"))
```

prints
```
tensor(True)
tensor(True)
tensor(True)
tensor(True)
```

One thing that was weird- `upsample_bilinear2d` and `upsample_trilinear3d` were only accurate across cpu/cuda with an epsilon of `1e-4`. That tentatively sounds close enough to say that cuda isn't "wrong" (?), but that's not exactly "equal"... and I also ran the script before my change, and `bilinear2d` and `trilinear3d` were also the same across cpu/cuda with an epsilon of `1e-4`.


Differential Revision: [D27351393](https://our.internmc.facebook.com/intern/diff/D27351393)

[ghstack-poisoned]

bdhirsh commented:
@ngimel Took out the bilinear/trilinear changes and added explicit tests. I confirmed that the tests fail on master and work after this change, so the inputs are "complicated enough" to hit this case.

ngimel commented on test/test_nn.py:
```diff
@@ -9247,6 +9247,13 @@ def test_upsamplingNearest2d(self):
+            # https://github.com/pytorch/pytorch/issues/54590
+            a = torch.arange(48).reshape(2, 2, 3, 4).contiguous(memory_format=torch.channels_last).float()
+            out_cpu = torch.nn.functional.interpolate(a, scale_factor=2, mode='nearest')
+            out_cuda = torch.nn.functional.interpolate(a.to('cuda'), scale_factor=2, mode='nearest')
```
this will fail if cuda is not available?

ngimel reviewed (commented):
You are fixing (and testing) only forward, I suspect the same thing is happening in backward, can you please add a test and fix if necessary?

ezyang reviewed (approved):
Perf is gonna be worse but I think this is for the best.

ezyang commented:
Plz address ngimel's comments

bdhirsh committed:
Update on "fix silent correctness bug with channels_last usage of upsample cuda kernels"

Fixes https://github.com/pytorch/pytorch/issues/54590

After the porting the upsample operators to be structured, they now forward memory_format information to the output. This is a problem for the cuda kernels, which are not implemented to deal with `torch.channels_last` memory format. After the structured port, the kernels now expect the output to already be made contiguous if it was not explicitly passed in. The cuda kernel is written to only handle "normal" contiguous tensors, and not channels_last contiguous tensors. The updated operators are:
* upsample_nearest2d
* upsample_nearest3d

It looks like upsample_bilinear2d and upsample_trilinear3d don't have this issue, so I didn't change those kernels.

This fix just allocates a temporary, contiguous output tensor when that happens, writes the results to the temporary and copies the results back to the output tensor.


Also added tests for upsample_nearest2d/3d to explicitly check that cuda and cpu results are the same. Those tests need to have input tensors with a mix of values and a large enough dimension size to distinguish between contiguous and channels_last behavior.


Differential Revision: [D27351393](https://our.internmc.facebook.com/intern/diff/D27351393)

[ghstack-poisoned]

bdhirsh commented:
@ngimel comments addressed, thanks. For the test, I just added a `cuda.is_available()` check.

For the backwards calls: `upsample_nearest2d/3d_backward` don't currently have channels_last logic in either cpu or cuda, so their meta functions currently don't forward memory_format information. (actually, none of the backwards ops do- I only added the memory_format-forwarding logic to the forward kernels in this [PR](https://github.com/pytorch/pytorch/pull/53535)). A test is definitely necessary though- I added calls to gradcheck/gradgradcheck in the original test. Let me know if that makes sense!

ezyang commented:
This is fine for now, though I suspect that backwards should also forward memory layout (and people just didn't get around to doing it)
